### PR TITLE
resample raw data

### PR DESCRIPTION
**Pull request by EmmaRenauld**
Last week we removed this function. That's because I tought it was resampling directly the SH, which is now straightforward with scilpy's upcoming function, you just have to give a different sphere.

But this function actually resamples RAW data, but uses the SH for the resampling. My bad. So I'm adding this back. I didn't check the last maths (last two lines). Can you confirm they are correct?

Thanks!

### CODE REVIEW

**Comment by AntoineTheb** [comment]
I might not be understanding the code well, but to me this looks like it resamples the SH, not the raw signal ?

File: dwi_ml/data/processing/dwi/dwi.py [hunk 1]
```diff
@@ -63,3 +62,52 @@ def standardize_data(data: np.ndarray, mask: np.ndarray = None,
...
+        Lambda-regularization coefficient in the SH fit, by default 0.006.
+
+    Returns
+    -------
+    resampled_dwi : np.ndarray (4D)
+        Resampled "raw" diffusion signal.
+    """
+    # Get SH fit
+    data_sh = compute_sh_coefficients(dwi_image, gradient_table, sh_order,
+                                      basis_type='tournier07', smooth=smooth)
```
Comment (mdesco) [inline dwi_ml/data/processing/dwi/dwi.py]: I would make basis_type an argument to the function, not hard-coded here... This is dangerous. 

And often, the SH basis will be descoteaux07

File: dwi_ml/data/processing/dwi/dwi.py [hunk 2]
```diff
@@ -63,3 +62,52 @@ def standardize_data(data: np.ndarray, mask: np.ndarray = None,
...
+    else:
+        sphere = get_sphere("repulsion100")
+    sh_basis = sph_harm_lookup.get("tournier07")
+
+    # Resample data
+    # B : 2-D array; real harmonics sampled at (\theta, \phi)
+    # m : array; degree of the sampled harmonics.
+    # l : array; order of the sampled harmonics.
+    B, m, l = sh_basis(sh_order, sphere.theta, sphere.phi)
+    data_resampled = np.dot(data_sh, B.T)
```
Comment (mdesco) [inline dwi_ml/data/processing/dwi/dwi.py]: This is correct. B.T contains the new sampling scheme and  B * SH projects to the sphere.

**Comment by EmmaRenauld** [comment]
Thanks Max. @ppoulin91 I am letting you decide how to answer to Maxime's comment that is is dangerous. Were you still using this?

I don't quite know how to test this, also.

**Comment by ppoulin91** [comment]
There is no response function involved, SH are fitted to the raw signal, then new directions are sampled using the provided sphere. 
@mdesco I don't think I understand your comment, can you clarify what you meant?

Sidenote: @AntoineTheb, in my understanding, the raw signal is composed of values on the sphere, so it's a (discretized) Spherical Function (because it's a single-shell signal). We use SH to go into a continuous space, then sample in another SF "space" (new directions on the sphere).

**Comment by mdesco** [comment]
@ppoulin91 I'm not talking about response functions. I'm talking about the SH basis and deciding to fix tournier versus descoteaux.  My initial point was that:
1) everywhere the default is descoteaux, so maybe safest to keep it as this
2) descoteaux basis is better conditioned than Tournier's basis (mathy/geaky discussion in my thesis).

All that to say, I would recommend using Descoteaux and initially, I thought you could receive sh coefficients but, in fact, this is my mistake. You are only using DWI signal on the sphere... so, using descoteaux or tournier would be identical in theory. But not, in practice. I would use Descoteaux. (I'm biased but that's ok)

**Comment by ppoulin91** [comment]
Ah ok, I get it. Yeah, I remember choosing tournier just for vizu in mrtrix at the time, and didn't change it back. Had forgotten about the better conditioning, so it would be better to go back to descoteaux and keep it that way.

**Pull request merged**